asset_dashboard/static/js/components/map_utils/MapClipper.js

```diff
@@ -0,0 +1,107 @@
+import React, { useEffect, useState } from 'react'
+import { FeatureGroup } from 'react-leaflet'
+import clip from 'turf-clip'
+import EditControl from './EditControl'
+
+export default function MapClipper({ geoJson, onClipped }) {
+  const [drawnGeometries, setDrawnGeometries] = useState(null)
+
+  useEffect(() => {
+    clipGeometries(drawnGeometries)
+  }, [drawnGeometries])
+
+  function onEdited(e) {
+    const layerKeys = getLayersKeys(e.layers._layers)
+
+    let editedLayers = {}
+
+    layerKeys.forEach(layerKey => {
+      const layer = e.layers._layers[layerKey]
+
+      const geoJSON = layer.toGeoJSON()
+
+      editedLayers = {
+        ...editedLayers,
+        [layerKey]: geoJSON
+      }
+    })
+
+    setDrawnGeometries(prevState => {
+      return {
+        ...prevState,
+        ...editedLayers
+      }
+    })
+  }
+
+  function onCreated(e) {
+    const geoJSON = e.layer.toGeoJSON()
+
+    setDrawnGeometries(prevState => {
+      return {
+        ...prevState,
+        [e.layer._leaflet_id]: geoJSON
+      }
+    })
+  }
+
+  function onDeleted(e) {
+    const layersKeys = getLayersKeys(e.layers._layers)
+
+    setDrawnGeometries(prevState => {
+      layersKeys.forEach(key => {
+        delete prevState[key]
+      })
+      return {
+        ...prevState
+      }
+    })
+  }
+
+  function getLayersKeys(layers) {
+    return Object.keys(layers).map(key => { return key })
+  }
+
+  function clipGeometries(geometries) {
+    if (geometries) {
+      const bounds = {
+        'type': 'FeatureCollection',
+        'features': Object.values(geometries)
+      }
+  
+      const clippedFeatureCollection = clip(bounds, geoJson)
+  
+      if (clippedFeatureCollection.features.length > 0) {
+        onClipped(clippedFeatureCollection)
+      }
```

Comment:
In what situation are there geometries but no features in the clipped feature collection?

Reply:
if a user draws a geometry but there aren't any underlying geometries.

the `clipGeometries` function will run anytime the `drawnGeometries` state changes, so whenever a user draws, edits, or deletes a polygon/rectangle with the draw tool. if they create a polygon that doesn't intersect with any of the geojson features, then `clippedFeatureCollection.features.length == 0`. in that case, we don't want to call the `onClipped` callback. adding this `clippedFeatureCollection.features.length > 0` is a truthy check so that `onClipped` doesn't get called, otherwise it will send invalid data that results in a 500 error on the server.

Reply:
upon further thought i think this assumption was based on the first iteration of the backend. in the case of the `features` array being empty, we will submit an empty featurecollection to the api and save it in the database (assuming the api would accept that, which i'm guessing would but i haven't tested). so i think we should still protect the backend from accepting an empty feature collection so we don't have a bunch of empty feature geometries stored in the database (assuming it would save an empty geometry...do you know if that is possible?)

Reply:
Sounds good. FWIW, I wasn't requesting any changes, I just wanted to understand how this condition is created.